indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java

```diff
@@ -0,0 +1,159 @@
+  }
+
+  public static IncludeExcludeFilter fromConfiguration() {
+    checkState(Configuration.isInitialized());
+    Set<String> filterProperties = Configuration.getConfig()
+        .stringPropertyNames()
+        .stream()
+        .filter(key -> key.startsWith(FILTER_CONFIG_PREFIX))
+        .collect(Collectors.toSet());
+    if (filterProperties.isEmpty()) {
+      return new IncludeExcludeFilter(ImmutableList.of(), ImmutableList.of());
+    }
+    List<Rule<String>> includeRules = new ArrayList<IncludeExcludeFilter.Rule<String>>();
+    List<Rule<String>> excludeRules = new ArrayList<IncludeExcludeFilter.Rule<String>>();
+    for (String propertyName : filterProperties) {
+      String propertyValue = Configuration.getString(propertyName, null).get();
+      logger.log(Level.FINEST, "Processing include/exclude rule {0}: {1}",
+          new Object[] { propertyName, propertyValue });
+      try {
+        if (propertyName.startsWith(INCLUDE_RULE_PREFIX)) {
+          includeRules.add(new Rule<>(new RegexPredicate(propertyValue)));
+        } else if (propertyName.startsWith(EXCLUDE_RULE_PREFIX)) {
+          excludeRules.add(new Rule<>(new RegexPredicate(propertyValue)));
+        }
+      } catch (PatternSyntaxException e) {
+        throw new InvalidConfigurationException("Invalid regex pattern " + propertyValue, e);
+      }
+    }
+    return new IncludeExcludeFilter(includeRules, excludeRules);
+  }
+
+  /**
+   * Returns true if the given string is included based on the configured include/exclude
+   * patterns.
+   *
+   * @param value a value to test
+   * @return true if the value is included based on the configuration
+   */
+  public boolean isAllowed(String value) {
+    boolean exclude = evaluateRules(excludeRules, value, false /* nothing is excluded */);
+    if (exclude) {
+      logger.log(Level.FINEST, "excluding " + value);
+      return false;
+    }
+    return evaluateRules(includeRules, value, true /* everything is included */);
+  }
+
+  private boolean evaluateRules(List<Rule<String>> rules, String value, boolean emptyRulesOutcome) {
+    if (rules.isEmpty()) {
+      return emptyRulesOutcome;
+    }
+    return rules.stream().map(r -> r.eval(value)).filter(e -> e == true).findFirst().orElse(false);
+  }
+
+  private static class Rule<T> {
+    private final Predicate<T> predicate;
+
+    public Rule(Predicate<T> predicate) {
+      this.predicate = checkNotNull(predicate);
+    }
+
+    public boolean eval(T val) {
+      return predicate.apply(val);
+    }
+  }
+
+  private static class RegexPredicate implements Predicate<String> {
+    private final Pattern p;
+    private final String regex;
+
+    private RegexPredicate(String regex) {
+      this.p = Pattern.compile(regex, Pattern.CASE_INSENSITIVE);
+      this.regex = regex;
+    }
+
+    @Override
+    public boolean apply(@Nullable String input) {
+      Matcher matcher = p.matcher(input);
+      boolean matches = matcher.matches();
```

Comment:
Could be combined with the previous line.

What do you think about matches vs find? The former could be unintuitive, and people will often not get any matches, whereas the latter might be prone to false positives. We could config as "...matches.foo = " or "...find.foo = ", rather than "...regex.foo = ", but I'm not too fond of the names in that context. I don't know if people are more likely to grok adding .* as needed, or ^ and $. We need to document this behavior, whatever we decide on.